File: Sourced Data/simplify_cmas_data.js
#!/usr/bin/env node

/**
 * Simplify CMAS data for the Colorado District Map App
 * This Node.js version doesn't require pandas
 * 
 * First install xlsx package:
 * npm install xlsx
 */

const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

function processWithJSON(data, outputFile) {
    // Filter for district-level, all-grades data
    const filtered = data.filter(row => 
        row['Level'] === 'DISTRICT' && 
        row['Grade'] === 'All Grades'
    );
    
    console.log(`Filtered to ${filtered.length} district records`);
    
    // Transform to desired format
    const result = {};
    
    filtered.forEach(row => {
        const district = row['District Name'] || row['DistrictName'] || row['Organization Name'];
        const subject = row['Subject'] || row['Content'];
        const year = (row['School Year'] || row['Year'] || '').toString().substring(0, 4);
        const percent = parseFloat(row['Percent Met or Exceeded Expectations'] || row['% Met or Exceeded'] || 0);
        
        if (!district || !year) return;
        
        if (!result[district]) {
            result[district] = { yearlyData: {} };
        }
        
        if (!result[district].yearlyData[year]) {
            result[district].yearlyData[year] = { subjects: [], total: 0 };
        }
        
        result[district].yearlyData[year].subjects.push(percent);
        result[district].yearlyData[year].total += percent;
    });
    
    // Calculate averages and format output
    Object.keys(result).forEach(district => {
        const cmasScores = [];
        const yearlyData = result[district].yearlyData;
        
        Object.keys(yearlyData).sort().forEach(year => {
            const avgScore = yearlyData[year].total / yearlyData[year].subjects.length;
            cmasScores.push({
                year: year,
                met_or_exceeded_pct: Math.round(avgScore * 10) / 10
            });
        });
        
        result[district] = {
            cmas_scores: cmasScores
        };
    });
    
    // Save to JSON
    fs.writeFileSync(outputFile, JSON.stringify(result, null, 2));
    
    console.log(`Saved simplified data to ${outputFile}`);
    console.log(`Processed ${Object.keys(result).length} districts`);
    
    // Show sample
    const sampleDistrict = Object.keys(result)[0];
    if (sampleDistrict) {
        console.log(`\nSample data for ${sampleDistrict}:`);
        console.log(JSON.stringify(result[sampleDistrict], null, 2));
    }
}

function simplifyCMASData(inputFile, outputFile) {
    console.log(`Reading ${inputFile}...`);
    
    try {
        // Read the Excel file
        const workbook = XLSX.readFile(inputFile);
        const sheetName = workbook.SheetNames[0];
        const worksheet = workbook.Sheets[sheetName];
        
        // Convert to JSON with header detection
        const rawData = XLSX.utils.sheet_to_json(worksheet, { header: 1 });
        
        // Find header row (look for row containing 'Level')
        let headerRowIndex = 0;
        for (let i = 0; i < Math.min(rawData.length, 40); i++) {
            if (rawData[i] && rawData[i].some(cell => cell && cell.toString() === 'Level')) {
                headerRowIndex = i;
                console.log(`Found header row at index ${i}`);
                break;
            }
        }
        
        // If no header found, try to read with sheet_to_json default behavior
        if (headerRowIndex === 0 && rawData.length > 0) {
            console.log('Trying alternative parsing method...');
            const jsonData = XLSX.utils.sheet_to_json(worksheet);
            if (jsonData.length > 0) {
                console.log('Sample row:', Object.keys(jsonData[0]));
                return processWithJSON(jsonData, outputFile);
            }
        }
        
        // Get headers and data
        const headers = rawData[headerRowIndex];
        const data = rawData.slice(headerRowIndex + 1);
        
        console.log('Found columns:', headers);
        
        // Find column indices
        const levelIdx = headers.findIndex(h => h && h.toString() === 'Level');
        const gradeIdx = headers.findIndex(h => h && h.toString() === 'Grade');
        const districtIdx = headers.findIndex(h => h && h.toString() === 'District Name');
        const subjectIdx = headers.findIndex(h => h && h.toString() === 'Content');
        const pctIdx = headers.findIndex(h => h && h.toString() === '2024'); // The 2024 column contains the percent met/exceeded
        
        // Filter for district-level, all-grades data
        const filtered = data.filter(row => 
            row[levelIdx] === 'DISTRICT' && 
            row[gradeIdx] === 'All Grades'
        );
        
        console.log(`Filtered to ${filtered.length} district records`);
        
        // Transform to desired format
        const result = {};
        
        filtered.forEach(row => {
            const district = row[districtIdx];
            const subject = row[subjectIdx];
            const year = '2024'; // This is 2024 data
            const percent = parseFloat(row[pctIdx]) || 0;
            
            if (!result[district]) {
                result[district] = { yearlyData: {} };
            }
            
            if (!result[district].yearlyData[year]) {
                result[district].yearlyData[year] = { subjects: [], total: 0 };
            }
            
            result[district].yearlyData[year].subjects.push(percent);
            result[district].yearlyData[year].total += percent;
        });
        
        // Calculate averages and format output
        Object.keys(result).forEach(district => {
            const cmasScores = [];
            const yearlyData = result[district].yearlyData;
            
            Object.keys(yearlyData).sort().forEach(year => {
                const avgScore = yearlyData[year].total / yearlyData[year].subjects.length;
                cmasScores.push({
                    year: year,
                    met_or_exceeded_pct: Math.round(avgScore * 10) / 10
                });
            });
            
            result[district] = {
                cmas_scores: cmasScores
            };
        });
        
        // Save to JSON
        fs.writeFileSync(outputFile, JSON.stringify(result, null, 2));
        
        console.log(`Saved simplified data to ${outputFile}`);
        console.log(`Processed ${Object.keys(result).length} districts`);
        
        // Show sample
        const sampleDistrict = Object.keys(result)[0];
        console.log(`\nSample data for ${sampleDistrict}:`);
        console.log(JSON.stringify(result[sampleDistrict], null, 2));
        
    } catch (error) {
        console.error('Error processing file:', error.message);
        console.error('Make sure to install xlsx: npm install xlsx');
        process.exit(1);
    }
}

// Main execution
if (require.main === module) {
    const inputFile = path.join(__dirname, 'General CMAS Score Data', '2024 CMAS ELA and Math District and School Summary Results.xlsx');
    const outputFile = path.join(__dirname, 'cmas_district_simplified.json');
    
    simplifyCMASData(inputFile, outputFile);
}

module.exports = { simplifyCMASData };

File: Sourced Data/simplify_cmas_data.py
#!/usr/bin/env python3
"""
Simplify CMAS data for the Colorado District Map App
Extracts only district-level, all-grades data with key metrics
"""

import pandas as pd
import json
import sys

def simplify_cmas_data(input_file, output_file):
    """
    Extract and simplify CMAS data for the mapping application.
    
    Filters for:
    - Level = 'DISTRICT' 
    - Grade = 'All Grades'
    
    Extracts columns:
    - District Name
    - Subject (ELA/Math)
    - Year
    - Percent Met or Exceeded
    - Number of Valid Scores (optional)
    """
    
    try:
        # Read Excel file
        print(f"Reading {input_file}...")
        
        # Try to auto-detect header row
        df_preview = pd.read_excel(input_file, nrows=20)
        
        # If that doesn't work, try reading with no header and find it manually
        if 'Level' not in df_preview.columns:
            df_raw = pd.read_excel(input_file, header=None, nrows=50)
            
            # Find header row by looking for 'Level' column
            header_row = 0
            for i in range(len(df_raw)):
                if 'Level' in str(df_raw.iloc[i].values):
                    header_row = i
                    break
            
            df = pd.read_excel(input_file, header=header_row)
        else:
            df = pd.read_excel(input_file)
        
        print(f"Found columns: {list(df.columns)}")
        
        # Filter for district-level, all-grades data
        filtered = df[
            (df['Level'] == 'DISTRICT') & 
            (df['Grade'] == 'All Grades')
        ].copy()
        
        print(f"Filtered to {len(filtered)} district records")
        
        # Identify key columns (common variations)
        district_col = next((col for col in df.columns if 'District' in col and 'Name' in col), 'District Name')
        subject_col = next((col for col in df.columns if 'Subject' in col), 'Subject')
        year_col = next((col for col in df.columns if 'Year' in col), 'School Year')
        
        # Find percent met/exceeded column
        pct_cols = [col for col in df.columns if 'Percent' in col and ('Met' in col or 'Exceeded' in col)]
        pct_col = pct_cols[0] if pct_cols else 'Percent Met or Exceeded Expectations'
        
        # Extract simplified data
        simplified = filtered[[district_col, subject_col, year_col, pct_col]].copy()
        simplified.columns = ['district', 'subject', 'year', 'percent_met_exceeded']
        
        # Clean year (extract just the year part if it's like "2023-2024")
        simplified['year'] = simplified['year'].astype(str).str[:4]
        
        # Convert to format matching the app's expected structure
        result = {}
        
        for district in simplified['district'].unique():
            district_data = simplified[simplified['district'] == district]
            
            # Group by year and calculate average across subjects
            yearly_avg = district_data.groupby('year')['percent_met_exceeded'].mean().round(1)
            
            # Create CMAS scores array
            cmas_scores = []
            for year, pct in yearly_avg.items():
                cmas_scores.append({
                    "year": year,
                    "met_or_exceeded_pct": float(pct)
                })
            
            result[district] = {
                "cmas_scores": sorted(cmas_scores, key=lambda x: x['year'])
            }
        
        # Save to JSON
        with open(output_file, 'w') as f:
            json.dump(result, f, indent=2)
        
        print(f"Saved simplified data to {output_file}")
        print(f"Processed {len(result)} districts")
        
        # Show sample
        sample_district = list(result.keys())[0]
        print(f"\nSample data for {sample_district}:")
        print(json.dumps(result[sample_district], indent=2))
        
    except Exception as e:
        print(f"Error processing file: {e}")
        print("Make sure pandas is installed: pip install pandas openpyxl")
        sys.exit(1)

if __name__ == "__main__":
    input_file = "General CMAS Score Data/2024 CMAS ELA and Math District and School Summary Results.xlsx"
    output_file = "cmas_district_simplified.json"
    
    simplify_cmas_data(input_file, output_file)